nixos/modules/services/web-apps/vikunja.nix

```diff
@@ -0,0 +1,167 @@
+    };
+    stateDir = mkOption {
+      type = types.str;
+      default = "/var/lib/vikunja";
+      description = "Vikunja data directory";
+    };
+    environmentFiles = mkOption {
+      type = types.listOf types.path;
+      default = [ ];
+      description = ''
+        List of environment files set in the vikunja systemd service.
+        For example passwords should be set in one of these files.
+      '';
+    };
+    setupNginx = mkOption {
+      type = types.bool;
+      default = config.services.nginx.enable;
+      defaultText = "config.services.nginx.enable";
+      description = "Whether to setup NGINX.";
+    };
+    frontendScheme = mkOption {
+      type = types.enum [ "http" "https" ];
+      description = ''
+        Whether the site is available via http or https.
+        This does not configure https or ACME in nginx!
+      '';
+    };
+    frontendHostname = mkOption {
+      type = types.str;
+      description = "The Hostname under which the frontend is running.";
+    };
+
+    settings = mkOption {
+      type = format.type;
+      default = {};
+      description = ''
+        Vikunja configuration. Refer to
+        <link xlink:href="https://vikunja.io/docs/config-options/"/>
+        for details on supported values.
+        '';
+    };
+    database = {
+      type = mkOption {
+        type = types.enum [ "sqlite" "mysql" "postgres" ];
+        example = "postgres";
+        default = "sqlite";
+        description = "Database engine to use.";
+      };
+      host = mkOption {
+        type = types.str;
+        default = "localhost";
+        description = "Database host address. Can also be a socket.";
+      };
+      user = mkOption {
+        type = types.str;
+        default = "vikunja";
+        description = "Database user.";
+      };
+      database = mkOption {
+        type = types.str;
+        default = "vikunja";
+        description = "Database name.";
+      };
+      path = mkOption {
+        type = types.str;
+        default = "${cfg.stateDir}/vikunja.db";
+        description = "Path to the sqlite3 database file.";
+      };
+      createDatabase = mkOption {
+        type = types.bool;
+        default = true;
+        description = "Whether to create a local database automatically.";
+      };
+    };
+  };
+  config = lib.mkIf cfg.enable {
+    services.vikunja.settings = {
+      database = {
+        inherit (cfg.database) type host user database path;
+      };
+      service = {
+        frontendurl = "${cfg.frontendScheme}://${cfg.frontendHostname}";
+      };
+    };
+
+    systemd.services.vikunja-api = {
+      description = "vikunja-api";
+      after = [ "network.target" ] ++ lib.optional usePostgresql "postgresql.service" ++ lib.optional useMysql "mysql.service";
+      wantedBy = [ "multi-user.target" ];
+      path = [ cfg.package-api ];
+      restartTriggers = [ configFile ];
+
+      serviceConfig = {
+        Type = "simple";
+        DynamicUser = true;
```

Comment:
I'm not sure if I'm understanding systemd's DynamicUser's well enough, but does this mean the service is executed with a different uid every time it starts? Because if that's the case that would break Vikunja's access to uploaded files.

Reply:
The first part is correct. The second not really (if I understood it correctly). The files are stored in `/var/lib/private/...` and symlinked to `/var/lib/vikunja` (the `StateDirectory`) with ownership to the dynamic user. So systemd manages that the StateDirectory is owned by the dynamic user. More details over [here](https://noise.getoto.net/2017/10/06/dynamic-users-with-systemd/).

Reply:
Okay, that makes a lot of sense then. Thanks for clarifying!